src/cudadev/DataFormats/SoALayout.h

```diff
@@ -0,0 +1,373 @@
+      CPP_TYPE::Scalar const* BOOST_PP_CAT(addressOf_, NAME)() const {                                               \
+        return parent_.BOOST_PP_CAT(NAME, _);                                                                        \
+      }                                                                                                              \
+      CPP_TYPE::Scalar* BOOST_PP_CAT(addressOf_, NAME)() { return parent_.BOOST_PP_CAT(NAME, _); }                   \
+)
+// clang-format on
+#define _DEFINE_METADATA_MEMBERS(R, DATA, TYPE_NAME) _DEFINE_METADATA_MEMBERS_IMPL TYPE_NAME
+
+/**
+ * Member assignment for trivial constructor
+ */
+#define _DECLARE_MEMBER_TRIVIAL_CONSTRUCTION_IMPL(VALUE_TYPE, CPP_TYPE, NAME) \
+  _SWITCH_ON_TYPE(VALUE_TYPE,                       /* Scalar */              \
+                  (BOOST_PP_CAT(NAME, _)(nullptr)), /* Column */              \
+                  (BOOST_PP_CAT(NAME, _)(nullptr)), /* Eigen column */        \
+                  (BOOST_PP_CAT(NAME, _)(nullptr))(BOOST_PP_CAT(NAME, Stride_)(0)))
+
+#define _DECLARE_MEMBER_TRIVIAL_CONSTRUCTION(R, DATA, TYPE_NAME) \
+  BOOST_PP_EXPAND(_DECLARE_MEMBER_TRIVIAL_CONSTRUCTION_IMPL TYPE_NAME)
+/**
+ * Computation of the column or scalar pointer location in the memory layout (at SoA construction time)
+ */
+#define _ASSIGN_SOA_COLUMN_OR_SCALAR_IMPL(VALUE_TYPE, CPP_TYPE, NAME)                                                 \
+  _SWITCH_ON_TYPE(VALUE_TYPE, /* Scalar */                                                                            \
+                  BOOST_PP_CAT(NAME, _) = reinterpret_cast<CPP_TYPE*>(curMem);                                        \
+                  curMem += (((sizeof(CPP_TYPE) - 1) / byteAlignment) + 1) * byteAlignment;                           \
+                  , /* Column */                                                                                      \
+                  BOOST_PP_CAT(NAME, _) = reinterpret_cast<CPP_TYPE*>(curMem);                                        \
+                  curMem += (((nElements_ * sizeof(CPP_TYPE) - 1) / byteAlignment) + 1) * byteAlignment;              \
+                  , /* Eigen column */                                                                                \
+                  BOOST_PP_CAT(NAME, _) = reinterpret_cast<CPP_TYPE::Scalar*>(curMem);                                \
+                  curMem += (((nElements_ * sizeof(CPP_TYPE::Scalar) - 1) / byteAlignment) + 1) * byteAlignment *     \
+                            CPP_TYPE::RowsAtCompileTime * CPP_TYPE::ColsAtCompileTime;                                \
+                  BOOST_PP_CAT(NAME, Stride_) = (((nElements_ * sizeof(CPP_TYPE::Scalar) - 1) / byteAlignment) + 1) * \
+                                                byteAlignment / sizeof(CPP_TYPE::Scalar);)                            \
```

Comment:
what happens if `byteAlignment` is not a multiple of `sizeof(CPP_TYPE::Scalar)` ?
what happens if `byteAlignment` is less than `sizeof(CPP_TYPE::Scalar)` ?

Reply:
Indeed, bad things might happen. Usual values are powers of 2 and a typical byte alignment is bigger than a built-in scalar, but this is not guaranteed. I think asserting (compile-time) that both `sizeof(CPP_TYPE::Scalar)` and `byteAlignment` are powers of 2 and aligning the columns to the biggest of the two should forbid any unwanted scenario while remaining a reasonable constraint on the parameters.

Reply:
The alignment must always be a power of 2.
The size of a type can be anything, but c++ (almost) mandates that it is a multiple of its own alignment.

But "scalar" in this context does not necessarily mean a primitive type - it could also be an aggregate like class or an array (e.g. `char name[1024]`).
Then the stride becomes 0, which I don't think is the intended behaviour ?
If `sizeof(CPP_TYPE::Scalar)` is greater that the alignment, the stride should just be `nElements_ * sizeof(CPP_TYPE::Scalar)` .

Reply:
Actually, shouldn't the stride simply be `(((nElements_ * sizeof(CPP_TYPE::Scalar) - 1) / byteAlignment) + 1) * byteAlignment`, without the `/ sizeof(CPP_TYPE::Scalar)` part ?